### DISCUSSION TITLE
Async steps at build time

### DISCUSSION THREAD

**Discussion by Sacharified** [vercel/next.js#12097] [Help]
I want to do some async work to fetch some data which will be injected in to my runtimeConfig in  `next.config.js`. As far as I'm aware this  is unsupported, is there a recommended way of doing async work at build time?

For example:

```
// next.config.js

module.exports = async () => {
  const token = await getToken();
  return {
    serverRuntimeConfig: {
      someToken: token
    }
  }
}
```

**Comment by leerob** [comment]
You need to get a token at build time. Could you use `getStaticProps` instead which will also run at build time?

**Reply by Sacharified** [reply]
I'm using SSR so that wouldn't work. I could use `getServerSideProps` but that will also run on every page request which is not really ideal.

**Reply by stigkj** [reply]
Can you have a "dummy" page that has a getStaticProps?

**Comment by stigkj** [comment]
The easiest would maybe be to put something in your build script that creates the `next.config.js` dynamically before  `next dev`/`next build`? I.e. have a template for `next.config.js` which you could inject the fetched token into.

**Comment by stigkj** [comment]
Is it possible to have this token as an environment variable? Then you could just query for that runtime.

**Comment by BigSully** [comment]
The following code shows how we solve a similar issue.
**Pass the object store to next/config, and populate the store by external service.** 
```javascript
const store = {}

let p = (async ()=>{
  let token = await getToken();
  store.token=token
  store.updateTime= Date.now()
})()

module.exports = {
    serverRuntimeConfig: {
      store
    }
  }
```

You can even update the store periodically, and the updates will still reflect.

**Reply by smcelhinney** [reply]
I'd love to understand why this ONLY works if you return the async value as an object property - this *doesn't* work
```javascript
let store;

let p = (async () => {
    let token = await getToken();
    store = { token, updateTime: Date.now()}
})()

module.exports = {
    serverRuntimeConfig: {
        store
    }
}
```

Does anyone know why?

**Reply by stigkj** [reply]
Maybe because you bind the reference to the store in `serverRuntimeConfig` and when `p` is run it changes the reference, i.e. it does not update what the old reference points to.

**Comment by GeorgeTaveras1231** [comment]
Would be great if next supported an async function to generate the config. Both webpack and storybook support this and its useful at times. I'm running into a similar issue where I need the webpack configs to be dynamic, dependent on the state of some file in the file-system. I could use the `sync` fs functions to do everything `sync` but that's generally a bad practice; especially if you have to read multiple files.

**Comment by knitkode** [comment]
hi @Sacharified , have you perhaps tried https://github.com/sindresorhus/make-synchronous ?

**Comment by glassdimlygr** [comment]
Edit: ignore below in favor of comment https://github.com/vercel/next.js/discussions/12097#discussioncomment-2487825
> +1 for async functions in `serverRuntimeConfig`. One can create an "initializing" state for the `store` in https://github.com/vercel/next.js/discussions/12097#discussion-1702, and then have code wait to access it until it is fully init'ed, however, would be nice if serverRuntimeConfig could be a callable async function or just await if it given a Promise instead of an object.

**Comment by glassdimlygr** [comment]
This ticket should be closed. You can do async work at build time after next `12.0.10`: https://nextjs.org/docs/messages/promise-in-next-config

Here's what I did

`next.config.js`:
```
module.exports = async () => {
const nconf = await initNconfConfig();
...
  serverRuntimeConfig: {
    getServerConfig: curryGetConfig(nconf),
  },
}
```

Other file:

```
const initNconfConfig = async (envName = process.env.NODE_ENV || 'development') => {
    // do async things
    return nconf;
}
```
```
const curryGetConfigWithNconf = (thisNconf) => getConfig = (path, brand = null) => {
    // do sync things like nconf.get('value')
    return foundValue;
}

```

index.tsx (a page)
```
const getServerSideProps = async () => {
  const testConfig = getConfig().serverRuntimeConfig.getServerConfig("cache:host");
  console.log({testConfig});

```

**Reply by pranavmshinde** [reply]
Hi @glassdimlygr, Can you please share the code of your next.config.js? I am bit confused on how to write configs inside this async function. Following code is giving error. I want to call an API inside this async function

````
module.exports = async () => {
  const cred = await getCred(); 
    env: {
      NEXT_PUBLIC_ENV: cred,
    },
    assetPrefix: "/mc-react", // Content switch prefix added to asset routes i.e. mc-react/_next
    async rewrites() {
      return {
        beforeFiles: [
          {
            source: `/mc-react/_next/data/:path*`,
            destination: '/_next/data/:path*'
           },
        ]
      }
    }
}
````
